Make Lenin hand a stack-level object

ApproximateIdentity opened the pull request:
This means that the hand follows around on different cards as expected. It also grabs onto objects on the current card and leaves them on the current card if cards are switched (which sounds confusing, but is what you would expect).

This does not make the hand a world level object.

ApproximateIdentity commented:
Okay the last commit makes the hand world-level and it works in all the ways you would want _except_ that the hand is invisible on other stacks. But the hand is working on other stacks because it will correctly grab buttons on those stacks and move them around.

@dkrasner Maybe this is related to the object ordering thing you were talking about on the call?

dkrasner commented:
@ApproximateIdentity yep that's the reason

you can always see if that's the case in the dev tools inspector. And a nice feature of it is that you can reorder the DOM elements with a drag&drop right there

![image](https://user-images.githubusercontent.com/1154326/114865877-2c40f380-9df3-11eb-8b55-4d81316171ce.png)


 i just dragged that (last-child) stack to be ahead of the hand image and all was fine

do you want me to add a fix on this branch so that stacks always come in order?

dkrasner commented on js/objects/System.js:
```diff
@@ -846,6 +846,36 @@ System._commandHandlers['importWorld'] = function(sender, sourceUrl){
                         // Tell the world to update current, in case
                         // a new StackView was attached with current set.
                         document.querySelector('st-world').updateCurrentStack();
+                        // This reorders all children of world so that non-stacks are at the end.
```
this shouldn't need to happen on import

if we make sure stacks are always added in order, like cards, then this will be preserved on serialization (and hence de-serialization)

ApproximateIdentity commented:
I think that your version just plain doesn't work. I rebased this branch on master and it continues to work (though using my method that maybe we don't like). I also pushed a branch `thomas-lenin-hand-stack-daniel-version` here https://github.com/UnitedLexCorp/SimpleTalk/tree/thomas-lenin-hand-stack-daniel-version which is your commit also rebased on master. In other words, these two branches are the same except for the last commit.

Yours is not working for me. I follow these steps to reproduce:

1. Start with bootstrap.html
2. Turn on hand detection
3. Import world bootstrap.html again
4. Look at the new screen and it is not there

That works on my branch (i.e. the hand stays on top on all stacks), but not on yours. I ran those steps again and put a breakpoint on the code here where your last commit made the changes:

https://github.com/UnitedLexCorp/SimpleTalk/blob/thomas-lenin-hand-stack-daniel-version/js/objects/System.js#L465

So the function that you made your changes in is never being called by the process.

Maybe @darth-cheney can check and verify that he's seeing the same thing?

ApproximateIdentity commented:
@dkrasner I changed this to only care about `st-image` objects and nothing else. You cool with this?

dkrasner commented:
> @dkrasner I changed this to only care about `st-image` objects and nothing else. You cool with this?



> I think that your version just plain doesn't work. I rebased this branch on master and it continues to work (though using my method that maybe we don't like). I also pushed a branch `thomas-lenin-hand-stack-daniel-version` here https://github.com/UnitedLexCorp/SimpleTalk/tree/thomas-lenin-hand-stack-daniel-version which is your commit also rebased on master. In other words, these two branches are the same except for the last commit.
> 
> Yours is not working for me. I follow these steps to reproduce:
> 
>     1. Start with bootstrap.html
> 
>     2. Turn on hand detection
> 
>     3. Import world bootstrap.html again
> 
>     4. Look at the new screen and it is not there
> 
> 
> That works on my branch (i.e. the hand stays on top on all stacks), but not on yours. I ran those steps again and put a breakpoint on the code here where your last commit made the changes:
> 
> https://github.com/UnitedLexCorp/SimpleTalk/blob/thomas-lenin-hand-stack-daniel-version/js/objects/System.js#L465
> 
> So the function that you made your changes in is never being called by the process.
> 
> Maybe @darth-cheney can check and verify that he's seeing the same thing?

@ApproximateIdentity  yes b/c you are testing the import of an _already existing_ world, as opposed to adding parts to world

So this works exactly as intended. On the  `thomas-lenin-hand-stack-daniel-version` branch I just ran 

![image](https://user-images.githubusercontent.com/1154326/115392350-c08acc00-a1e0-11eb-9f00-2bd04972060d.png)


note the added stack comes before the image, even though the opposite order commands was run. 


Again this is nothing to do with importing worlds. Those should just come in as is. 


Perhaps the best thing to do is to stop the image recognition if a world is imported and then start it again. This is simple and allow us to see how things play out, what kinds of parts we'll want on world or anywhere else etc

ApproximateIdentity commented:
@dkrasner I changed it so that importing world stops and restarts hand recognition if it's running. i also added a second commit to do essentially the same thing when saving a world which should satisfy this issue: https://github.com/UnitedLexCorp/SimpleTalk/issues/187

thoughts about this method?

dkrasner commented:
@ApproximateIdentity i think this is fine for now. In the future all this card and stack priority will need to be solved with some sort of `slot`'ing   otherwise we'll end up chasing issues like this all over the place (as opposed insisting in one place where cards and stacks go - i'll work on this in the next little while)

ApproximateIdentity commented:
okay cool. yeah i think the only reason this is sort of acceptable for the moment is because the hand is now the only example of this. but the second we want to do something else similar, it definitely needs a less one-off solution.
